src/semidiscretization/semidiscretization_hyperbolic.jl

```diff
@@ -81,22 +81,47 @@ end
 
 
 # allow passing named tuples of BCs constructed in an arbitrary order
-digest_boundary_conditions(boundary_conditions) = boundary_conditions
+digest_boundary_conditions(boundary_conditions, cache) = boundary_conditions
 
-function digest_boundary_conditions(boundary_conditions::NamedTuple{Keys,ValueTypes}) where {Keys, ValueTypes<:NTuple{2,Any}} # 1D
+function digest_boundary_conditions(boundary_conditions::NamedTuple{Keys,ValueTypes}, cache) where {Keys, ValueTypes<:NTuple{2,Any}} # 1D
   @unpack x_neg, x_pos = boundary_conditions
   (; x_neg, x_pos)
 end
-function digest_boundary_conditions(boundary_conditions::NamedTuple{Keys,ValueTypes}) where {Keys, ValueTypes<:NTuple{4,Any}} # 2D
+function digest_boundary_conditions(boundary_conditions::NamedTuple{Keys,ValueTypes}, cache) where {Keys, ValueTypes<:NTuple{4,Any}} # 2D
   @unpack x_neg, x_pos, y_neg, y_pos = boundary_conditions
   (; x_neg, x_pos, y_neg, y_pos)
 end
-function digest_boundary_conditions(boundary_conditions::NamedTuple{Keys,ValueTypes}) where {Keys, ValueTypes<:NTuple{6,Any}} # 3D
+function digest_boundary_conditions(boundary_conditions::NamedTuple{Keys,ValueTypes}, cache) where {Keys, ValueTypes<:NTuple{6,Any}} # 3D
   @unpack x_neg, x_pos, y_neg, y_pos, z_neg, z_pos = boundary_conditions
   (; x_neg, x_pos, y_neg, y_pos, z_neg, z_pos)
 end
 
-function digest_boundary_conditions(boundary_conditions::AbstractArray)
+# peel apart the boundary condition dictionary for the UnstructuredQuadMesh solver and sort the
+# boundary types and their associated boundary indices into tuples for use later
+function digest_boundary_conditions(boundary_conditions::Dict, cache)
+  # extract the unique boundary function routines from the dictionary
+  boundary_types = Tuple(unique(collect(values(boundary_conditions))))
+
+  # pull and sort the indexing for each boundary type
+  boundary_indices = Vector{Any}(nothing, length(boundary_types))
+  for j in 1:length(boundary_types)
+    indices_for_current_type = Int[]
+    for (test_name, test_condition) in boundary_conditions
+      temp_indices = findall(x->x===test_name, cache.boundaries.name)
+      if test_condition === boundary_types[j]
+        indices_for_current_type = vcat(indices_for_current_type, temp_indices)
+      end
+    end
+    boundary_indices[j] = sort(indices_for_current_type)
+  end
+
+  # put everything together into a NamedTuple that has tuples with the (sorted) boundary type and
+  # its associated indices. Save a copy of the original dictionary for printing
+  ( sorted_boundary_information = tuple(zip(boundary_types, boundary_indices)...),
+    boundary_dictionary = boundary_conditions, )
```

Comment:
The sorting you did above is already quite good, thanks :+1: However, i think it would be cleaner to create a new `struct` to hold the information you return here and use that for dispatch later. It could be something like
```julia
struct AndrewPicksSomeGreatNameHereSinceDontWantToDoThatRightNow{N, BCs<:NTuple{N, Any}}
  boundary_conditions::BCs
  boundary_indices::NTuple{N, Vector{Int}}
  boundary_dictionary::Dict{Symbol, Any}
end
```
Then, you can iterate over the tuples in a type-stable way using "lispy tuple programming" - search for that phrase in Trixi's source code, I'm sure I used it twice or so. That should remove the remaining allocations and make the `rhs!` stuff type stable and fast.

Reply:
This is a nice idea. It would also make the display `if` for the semidiscretization a bit cleaner. I just need to figure out where this struct would live. Would I need to make a new `semidiscretization` constructor? Or could I still pass this new struct as the general `boundary_conditions`argument?

Reply:
You can let the `struct` live in the quad mesh solver file and construct it in `digest_boundary_conditions` as you do now.

Reply:
Groovy, I will work on it. My naive way already reduced memory allocations a bit but hopefully this struct way will also reduce the actual computation time